sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java

```diff
@@ -0,0 +1,1107 @@
+    }
+
+    /**
+     * DoFn to fetch a resource from an Google Cloud Healthcare FHIR store based on resourceID
+     *
+     * <p>This DoFn consumes a {@link PCollection} of notifications {@link String}s from the FHIR
+     * store, and fetches the actual {@link String} object based on the id in the notification and
+     * will output a {@link PCollectionTuple} which contains the output and dead-letter {@link
+     * PCollection}*.
+     *
+     * <p>The {@link PCollectionTuple} output will contain the following {@link PCollection}:
+     *
+     * <ul>
+     *   <li>{@link FhirIO.Read#OUT} - Contains all {@link PCollection} records successfully read
+     *       from the Fhir store.
+     *   <li>{@link FhirIO.Read#DEAD_LETTER} - Contains all {@link PCollection} of {@link
+     *       HealthcareIOError}* of message IDs which failed to be fetched from the Fhir store, with
+     *       error message and stacktrace.
+     * </ul>
+     */
+    public static class FetchResourceJsonString
+        extends PTransform<PCollection<String>, FhirIO.Read.Result> {
+
+      /** Instantiates a new Fetch Fhir message DoFn. */
+      public FetchResourceJsonString() {}
+
+      @Override
+      public FhirIO.Read.Result expand(PCollection<String> resourceIds) {
+        return new FhirIO.Read.Result(
+            resourceIds.apply(
+                ParDo.of(new FetchResourceJsonString.StringGetFn())
+                    .withOutputTags(FhirIO.Read.OUT, TupleTagList.of(FhirIO.Read.DEAD_LETTER))));
+      }
+
+      /** DoFn for fetching messages from the Fhir store with error handling. */
+      public static class StringGetFn extends DoFn<String, String> {
+
+        private Counter failedMessageGets =
+            Metrics.counter(FetchResourceJsonString.StringGetFn.class, "failed-message-reads");
+        private static final Logger LOG =
+            LoggerFactory.getLogger(FetchResourceJsonString.StringGetFn.class);
+        private final Counter successfulStringGets =
+            Metrics.counter(
+                FetchResourceJsonString.StringGetFn.class, "successful-hl7v2-message-gets");
+        private HealthcareApiClient client;
+        private ObjectMapper mapper;
+
+        /** Instantiates a new Hl 7 v 2 message get fn. */
+        StringGetFn() {}
+
+        /**
+         * Instantiate healthcare client.
+         *
+         * @throws IOException the io exception
+         */
+        @Setup
+        public void instantiateHealthcareClient() throws IOException {
+          this.client = new HttpHealthcareApiClient();
+          this.mapper = new ObjectMapper();
+        }
+
+        /**
+         * Process element.
+         *
+         * @param context the context
+         */
+        @ProcessElement
+        public void processElement(ProcessContext context) {
+          String resourceId = context.element();
+          try {
+            context.output(fetchResource(this.client, resourceId));
```

Comment:
this function makes RPCs to an external service one-by-one. This will be slow. It may be worth buffering these requests, and applying them with a thread pool. You can add this optimization later on, though.

Reply:
The [Fhir.read](https://cloud.google.com/healthcare/docs/reference/rest/v1beta1/projects.locations.datasets.fhirStores.fhir/read) api doesn't support requesting several resources in a single call.

We could look into [Fhir.search](https://cloud.google.com/healthcare/docs/reference/rest/v1beta1/projects.locations.datasets.fhirStores.fhir/search) to request all resources with IDs in a list.

Reply:
You could perform multiple `Fhir.read` requests from separate threads? This may be fine for now, though, and if you find that the extra optimization will help, then add it later on?